src/codegen/encoding/borsh_encoding.rs

```diff
@@ -693,4 +717,584 @@ impl BorshEncoding {
+        struct_ty: &StructType,
+        validator: &mut BufferValidator,
+        ns: &Namespace,
+        vartab: &mut Vartable,
+        cfg: &mut ControlFlowGraph,
+    ) -> (Expression, Expression) {
+        let size = if let Some(no_padding_size) = ns.calculate_struct_non_padded_size(struct_ty) {
+            let padded_size = expr_ty.solana_storage_size(ns);
+            // If the size without padding equals the size with padding,
+            // we can memcpy this struct directly.
+            if padded_size.eq(&no_padding_size) {
+                let size = Expression::NumberLiteral(Loc::Codegen, Type::Uint(32), no_padding_size);
+                validator.validate_offset_plus_size(&offset, &size, vartab, cfg);
+                let source_address = Expression::AdvancePointer {
+                    pointer: Box::new(buffer.clone()),
+                    bytes_offset: Box::new(offset),
+                };
+                let allocated_struct = vartab.temp_anonymous(expr_ty);
+                cfg.add(
+                    vartab,
+                    Instr::Set {
+                        loc: Loc::Codegen,
+                        res: allocated_struct,
+                        expr: Expression::StructLiteral(Loc::Codegen, expr_ty.clone(), vec![]),
+                    },
+                );
+                let struct_var =
+                    Expression::Variable(Loc::Codegen, expr_ty.clone(), allocated_struct);
+                cfg.add(
+                    vartab,
+                    Instr::MemCopy {
+                        source: source_address,
+                        destination: struct_var.clone(),
+                        bytes: size.clone(),
+                    },
+                );
+                return (struct_var, size);
+            } else {
+                // This struct has a fixed size, but we cannot memcpy it due to
+                // its padding in memory
+                Some(Expression::NumberLiteral(
+                    Loc::Codegen,
+                    Type::Uint(32),
+                    no_padding_size,
+                ))
+            }
+        } else {
+            None
+        };
+
+        let struct_tys = struct_ty
+            .definition(ns)
+            .fields
+            .iter()
+            .map(|item| item.ty.clone())
+            .collect::<Vec<Type>>();
+
+        // If it was not possible to validate the struct beforehand, we validate each field
+        // during recursive calls to 'read_from_buffer'
+        let mut struct_validator = BufferValidator {
+            buffer_length: validator.buffer_length.clone(),
+            types: &struct_tys[..],
+            verified_until: if validator.validation_necessary() {
+                -1
+            } else {
+                struct_tys.len() as i32
+            },
+            current_arg: if validator.validation_necessary() {
+                0
+            } else {
+                struct_tys.len()
+            },
+        };
```

Comment:
I do really like how you've solved the problem of minimizing the amount of bounds checks. However, I think this construct here should really be a method on `BufferValidator`. Something like `fn subvalidator()`.